added wallet attestation scheme (#52)
vcstuff/oid4vc-haip-sd-jwt-vc: draft-oid4vc-haip-sd-jwt-vc.md

In addition to this definition, the Wallet Attestation MAY contain the following claims in the `cnf` element:

* `key_type`: OPTIONAL. JSON String that asserts the security mechanism the Wallet uses to manage the private key associated with the public key given in the `cnf` claim. This mechanism is based on the capabilities of the execution environent of the Wallet, this might be a secure element (in case of a wallet residing on a smartphone) or a Cloud-HSM (in case of a cloud Wallet). This specification defines the following values for `key_type`:
  * `software`: It MUST be used when the Wallet uses software-based key management.
  * `hardware`: It MUST be used when the wallet uses hardware-based key management.
  * `tee`: It SHOULD be used when the Wallet uses the Trusted Execution Environment for key management.
  * `secure_enclave`: It SHOULD be used when the Wallet uses the Secure Enclave for key management.
  * `strong_box`: It SHOULD be used when the Wallet uses the Strongbox for key management.
  * `secure_element`: It SHOULD be used when the Wallet uses a Secure Element for key management.
  * `hsm`: It SHOULD be used when the Wallet uses Hardware Security Module (HSM).

> * `user_authentication`: OPTIONAL. JSON String that asserts the security mechanism the Wallet uses to authenticate access to the private key associated with the public key given in the `cnf` claim. This specification defines the following values for `user_authentication`: `system_biometry`, `system_pin`, `internal_biometry`, `internal_pin`, and `secureelement_pin`.

Author: @paulbastian can you please provide definitions of the user authentication values?
@Sakurann I think we either don't define the claim at all or we fully define it.

> * `user_authentication`: OPTIONAL. JSON String that asserts the security mechanism the Wallet uses to authenticate access to the private key associated with the public key given in the `cnf` claim. This specification defines the following values for `user_authentication`: `system_biometry`, `system_pin`, `internal_biometry`, `internal_pin`, and `secureelement_pin`.

Reviewer: ```suggestion
* `user_authentication`: OPTIONAL. JSON String that asserts the security mechanism the Wallet uses to authenticate access to the private key associated with the public key given in the `cnf` claim. This specification defines the following values for `user_authentication`:
  * `system_biometry`: It MUST be used when the key usage is authorized through a biometric factor by the operating system.
  * `system_pin`: It MUST be used when the key usage is authorized through a knowledge factor by the operating system.
  * `internal_biometry`: It MUST be used when the key usage is authorized through a biometric factor by the Wallet itself.
  * `internal_pin`: It MUST be used when the key usage is authorized through a knowledge factor by the Wallet itself.
  * `secure_element_pin` It MUST be used when the key usage is authorized through a knowledge factor by the secure element that is managing the key itself.
```

Reviewer: We may use JSON Array for `user_authentication`?

> * `user_authentication`: OPTIONAL. JSON String that asserts the security mechanism the Wallet uses to authenticate access to the private key associated with the public key given in the `cnf` claim. This specification defines the following values for `user_authentication`: `system_biometry`, `system_pin`, `internal_biometry`, `internal_pin`, and `secureelement_pin`.

Author: @Sakurann Paul updated the PR with definitions. Please check.

  * `software`: It MUST be used when the Wallet uses software-based key management.
  * `hardware`: It MUST be used when the wallet uses hardware-based key management.
  * `tee`: It SHOULD be used when the Wallet uses the Trusted Execution Environment for key management.
  * `secure_enclave`: It SHOULD be used when the Wallet uses the Secure Enclave for key management.
  * `strong_box`: It SHOULD be used when the Wallet uses the Strongbox for key management.
  * `secure_element`: It SHOULD be used when the Wallet uses a Secure Element for key management.
  * `hsm`: It SHOULD be used when the Wallet uses Hardware Security Module (HSM).
* `user_authentication`: OPTIONAL. JSON String that asserts the security mechanism the Wallet uses to authenticate access to the private key associated with the public key given in the `cnf` claim. This specification defines the following values for `user_authentication`:

>   * `system_biometry`: It MUST be used when the key usage is authorized through a biometric factor by the operating system.
>   * `system_pin`: It MUST be used when the key usage is authorized through a knowledge factor by the operating system.

Reviewer 2: ```suggestion
  * `system_biometry`: It MUST be used when the key usage is authorized through a biometric factor by the mobile operating system.
  * `system_pin`: It MUST be used when the key usage is authorized through a knowledge factor by the mobile operating system.
```
there are different OSs

> * `user_authentication`: OPTIONAL. JSON String that asserts the security mechanism the Wallet uses to authenticate access to the private key associated with the public key given in the `cnf` claim. This specification defines the following values for `user_authentication`:
>   * `system_biometry`: It MUST be used when the key usage is authorized through a biometric factor by the operating system.
>   * `system_pin`: It MUST be used when the key usage is authorized through a knowledge factor by the operating system.
>   * `internal_biometry`: It MUST be used when the key usage is authorized through a biometric factor by the Wallet itself.
>   * `internal_pin`: It MUST be used when the key usage is authorized through a knowledge factor by the Wallet itself.
>   * `secure_element_pin` It MUST be used when the key usage is authorized through a knowledge factor by the secure element that is managing the key itself.

Reviewer 2: ```suggestion
* `user_authentication`: OPTIONAL. JSON String that asserts the security mechanism the Wallet uses to authenticate the user to authorize access to the private key associated with the public key given in the `cnf` claim. This specification defines the following values for `user_authentication`:
  * `system_biometry`: It MUST be used when the key usage is authorized by the mobile operating system using a biometric factor.
  * `system_pin`: It MUST be used when the key usage is authorized by the mobile operating system using personal identification number (PIN) as a knowledge factor.
  * `internal_biometry`: It MUST be used when the key usage is authorized by the Wallet using a biometric factor.
  * `internal_pin`: It MUST be used when the key usage is authorized by the Wallet using PIN as a knowledge factor.
  * `secure_element_pin` It MUST be used when the key usage is authorized by the secure element managing the key itself using PIN as a knowledge factor.
```

`knowledge factor` != `pin`, no? knowledge factor can be password, etc...
I have reservations by saying `by the Wallet`, given that the term `Wallet` is not defined in HAIP yet (opened #64), prefer something like `at the application layer`, but guess it's ok for now.

Reviewer: I used knowledge factor because for in case of operating system, Android can not distinguish between password, pin and swipe pattern. Secure element will always have a pin and usually the wallet option as well

Reviewer 2: what does that mean in terms of the current/proposed spec text?

Author: > I have reservations by saying by the Wallet
The section is about wallet attestation, so I think it is reasonable to use the term "Wallet"

Author: @Sakurann @paulbastian I suggest to just remove "as a knowledge factor" as I think PIN is sufficient.

Reviewer 2: > I suggest to just remove "as a knowledge factor" as I think PIN is sufficient.

I edited the suggestion above - please apply it if you are ok with it

Reviewer 2: > @Sakurann @paulbastian I suggest to just remove "as a knowledge factor" as I think PIN is sufficient.

sure, but the wallet needs to be defined - ideally in this PR, but for now I opened a separate issue #64

Reviewer: looks fine to me

The Wallet Attestation MAY also contain the following claim:

* `aal`: OPTIONAL. JSON String asserting the authentication level of the Wallet and the key as asserted in the `cnf` claim.

To obtain the issuer's Public key for verification, wallet attestions MUST support web-based key resolution as defined in Section 5 of [@!I-D.terbu-sd-jwt-vc]. The JOSE header `kid` MUST be used to identify the respective key.

This is an example of a Wallet Instance Attestation:

```json
{
  "typ": "wallet-attestation+jwt",
  "alg": "ES256",
  "kid": "1"
}
.
{
  "iss": "<identifier of the issuer of this wallet attestation>",
  "sub": "<`client_id` of the OAuth client>",
  "iat": 1516247022,
  "exp": 1541493724,
  "aal" : "https://trust-list.eu/aal/high",
  "cnf": {
    "jwk": {
      "kty": "EC",
      "crv": "P-256",
      "x": "TCAER19Zvu3OHF4j4W4vfSVoHIP1ILilDls7vCeGemc",
      "y": "ZxjiWWbZMQGHVWKVQ4hbSIirsVfuecCE6t4jT9F2HZQ"
    },
    "key_type": "STRONGBOX",

>     "user_authentication": "SYSTEM_PIN",

Reviewer 3: shouldn't these values be lower cased to match the definitions above?

## Token Endpoint {#token-endpoint}

   * The Wallets MUST perform client authentication as defined in [@!I-D.ietf-oauth-attestation-based-client-auth].
   * Refresh tokens MUST be supported for credential refresh.
   * Wallets MUST support deferred authorization by being able to process the Token error response parameters `authorization_pending` and `slow_down`, and the credential offer parameter `interval`.

>    * The wallet attestation JWT scheme is defined in (#wallet-attestation-schema).

Reviewer 4: ```suggestion
   * The Wallet Attestation JWT scheme is defined in (#wallet-attestation-schema).
```

### Wallet Attestation Schema {#wallet-attestation-schema}

> Wallets MUST use attestations following the definition given in [@!I-D.looker-oauth-attestation-based-client-auth].

Reviewer 4: I would be better to set it as SHOULD, since it is not normative HOW the attestation could be provisioned and obtained. The MUST should be considered for the verification of the attestation, not to the way it was provisioned

I say this because there are several ways to provision a signed JWT and I would keep the door open with a good guidance (SHOULD) without forcing the usage of a particular specification

> Wallets MUST use attestations following the definition given in [@!I-D.looker-oauth-attestation-based-client-auth].

Reviewer 5: Now I-D.ietf-oauth-attestation-based-client-auth

> Wallets MUST use attestations following the definition given in [@!I-D.looker-oauth-attestation-based-client-auth].

Author: I rewrote the paragraph and changed it to SHOULD as with the new revision we have options on different abstraction levels to convey information about the security of the wallet and the key.

> In addition, the Wallet Attestation MUST contain the following claims:

Reviewer 2: In addition to what? To the claims defined in draft-ietf-looker?

Author: In addition to "the definition given in [!I-D.ietf-looker-key-attestation-client-authentication]"

> * `key_type`: this claim asserts the security mechanism the wallet can use to manage private keys. This capability is based on the capabilities of the execution environent of the wallet, this might be a secure element (in case of a wallet residing on a smartphone) or a Cloud-HSM (in case of a cloud wallet). This specification defines the following values for `key_type`: `STRONGBOX`, `TEE`, `SecureEnclave`, `Software`.

Reviewer 2: are the values taken from somewhere or defined in HAIP? If the latter, why not use snake case?

> * `key_type`: this claim asserts the security mechanism the wallet can use to manage private keys. This capability is based on the capabilities of the execution environent of the wallet, this might be a secure element (in case of a wallet residing on a smartphone) or a Cloud-HSM (in case of a cloud wallet). This specification defines the following values for `key_type`: `Software`, `TEE`, `Strongbox`, `Secure Enclave`, and `Secure Element`.

Reviewer: ```suggestion
* `key_type`: this claim asserts the security mechanism the wallet can use to manage private keys. This capability is based on the capabilities of the execution environent of the wallet, this might be a secure element (in case of a wallet residing on a smartphone) or a Cloud-HSM (in case of a cloud wallet). This specification defines the following values for `key_type`: `Software`, `TEE`, `Strongbox`, `Secure Enclave`, `Secure Element` and `External-HSM`.
```

> * `key_type`: this claim asserts the security mechanism the wallet can use to manage private keys. This capability is based on the capabilities of the execution environent of the wallet, this might be a secure element (in case of a wallet residing on a smartphone) or a Cloud-HSM (in case of a cloud wallet). This specification defines the following values for `key_type`: `Software`, `TEE`, `Strongbox`, `Secure Enclave`, `Secure Element` and `External-HSM`.

Reviewer 5: I wouldn't include spaces in the values.  They can cause quirky problems, in practice.

Reviewer 4: I would propose to use snake_case for the key values, then

- software
- tee
- strongbox
- secure_enclave
- secure_element
- external-hsm

Author: modified

> * `user_authentication`: this claim asserts the security mechanism the wallet can use to authenticate access to private keys. This specification defines the following values for `user_authentication`: `APP_PIN_6_DIGITS`, ...

Reviewer 2: Again, where are the values being defined..? If not in HAIP should it point somewhere?

Author: I would prefer to reuse WebAuthn values. Let's discuss.

Reviewer 4: if WebAuthn values are used I'm ok with that, we then need a ref to them

> * `user_authentication`: this claim asserts the security mechanism the wallet can use to authenticate access to private keys. This specification defines the following values for `user_authentication`: `System-Biometry`, `System-PIN`, `Internal-Biometry`, `Internal-PIN`, and `SecureElement-PIN`.

Reviewer 4: these information increases the responsability to the issuer that should evaluate mixed variables to an qualitative index

I'm not in favor of this, since, as already made with ACR values (LoA) in SAML2 and OIDC, a trust framework just have to define the properties within which some attestation levels are achieved

that's why we're working with `attested_security_context` 

the regulation should map key types and user auth types, by quality, inside different level (low, middle, high?)

very often the technology changes, with upgrades and deprecation, using a normalized and indexed approach is better than expose issuer to internal evaluation of the security properties that, not at least, then exposes also personal tastes like the user auth type or differently, the key_Type that in some cases are mobile-brand specific -> then discloses the hardware type

Reviewer: as discussed in the call yesterday, as long as we don't have a complete framework on attested security context and the mappings of key type and userauthentication to asc, it makes sense to offer key_type and user_authentication as this is a mechanism that is already working today and has been implemented.
Once we have a complete trust framework with asc mappings, we can add them to the client attestations

Therefore, I'm voting to keep them in for now and offer both possiblities

Reviewer 2: In general I agree with Giuseppe on the direction, but what Paul is saying that we are lacking those "normalized values" is also true.. so if we foresee that concrete `key_type` and `user_authentication` will be replaced by "normalized approach", I think at least we should not mandate key_type and user_authentication?

Reviewer 2: JohnB suggested `external-multifactor`

> * `user_authentication`: this claim asserts the security mechanism the wallet can use to authenticate access to private keys. This specification defines the following values for `user_authentication`: `System-Biometry`, `System-PIN`, `Internal-Biometry`, `Internal-PIN`, and `SecureElement-PIN`.

Reviewer 2: ```suggestion
* `user_authentication`: JSON string of a claim that asserts the security mechanism the wallet can use to authenticate access to private keys. This specification defines the following values for `user_authentication`: `System-Biometry`, `System-PIN`, `Internal-Biometry`, `Internal-PIN`, and `SecureElement-PIN`.
```

Is it a string or array of strings??

Reviewer 2: I honestly do not know what any of the following values for `user_authentication` mean: `System-Biometry`, `System-PIN`, `Internal-Biometry`, `Internal-PIN`, and `SecureElement-PIN`." suggest we open an issue for now.

> These additional claims inform the issuer about the security capabilities of the wallet and allows the issuer to refuse credential issuance if the achievble security level of a certain wallet does not fulfil the issuer's requirements.

Reviewer 4: Do we have to defined error codes and messages for that?

Author: you mean in the credential issuance response?

Reviewer 4: yes, do we have to give a ref to openid4vc?

Reviewer 4: https://github.com/vcstuff/oid4vc-haip-sd-jwt-vc/issues/62

> Wallet attestions MUST support web-based key resolution as defined in Section 5 of [@!I-D.terbu-sd-jwt-vc]. The JOSE header `kid` MUST be used to identify the respective key.

Reviewer 2: ```suggestion
To obtain the issuer's Public key for verification, wallet attestions MUST support web-based key resolution as defined in Section 5 of [@!I-D.terbu-sd-jwt-vc]. The JOSE header `kid` MUST be used to identify the respective key.
```

This is an example of a wallet attestation:

```json
{
  "typ": "wallet-attestation+jwt",
  "alg": "ES256",
  "kid": "1"
}
.
{
  "iss": "https://wallet.example.com",

>   "sub": "https://wallet.example.com",

Reviewer 4: I'm not in favor of duplicating iss to sub

I have the strong assumption that the wallet provider and the wallet instance are different entities

In my current implementation we have decided to use the jwk thumbprint to identify the wallet instance and then we use this value in the sub

there's a good agreement to simply remove the sub claim, WDYT about this?

Author: > sub descibes the specific Wallet Solution of a Wallet Instance
which includes the option to leave instances anonymous, right?

Reviewer: yes, anonymous instances should be RECOMMENDED in my opinion

Reviewer 4: do we assume that a single wallet provider may own multiple wallet solution?

I assume that a single wallet provider owns a single wallet solution.
If we don't have agreement on how the sub value should be valued, considering that it may have the same value of iss and that's misleading since provider and instance are different entities, I suggest to get sub out by simply removing it from the required attributes

Reviewer 6: Is the **Wallet Provider** always the **Attester** of the **Wallet Instance**?

If that is not the case, it means there are three entities, the Wallet Provider, the Wallet Instance and the Attester. It is theoretically impossible to map the three entities to the two claims, `iss` and `sub`, without conflicts.

If the Wallet Provider is NOT always the Attester, the definition of the `iss` claim by eIDAS is problematic and conflicts with the specification of the attestation-based client authentication.

Author: According to the underlying draft-looker-oauth-attestation-based-client-auth, the `iss` is the entity the AS (here the wallet) trusts for the purpose of the issuance of attestations and the `sub` is the client id. The wallet can establish the trust be looking the `iss` value up in a trusted list, for example. The actual wallet product could be the `sub` value. The attributes (including capabilities and perhaps even authorization) need to be conveyed through claims in the attestation. 
The difference between attester and wallet provider would be that if the wallet provider is in the iss, it needs to be entitled by the trusted list to issue it's own attestations whereas in the case of the attester, it's the attester who has that entitlement. 

Does this sound reasonable?

Author: According to the underlying draft-looker-oauth-attestation-based-client-auth, the `iss` is the entity the AS (here the wallet) trusts for the purpose of the issuance of attestations and the `sub` is the client id. The wallet can establish the trust be looking the `iss` value up in a trusted list, for example. The actual wallet product could be the `sub` value. The attributes (including capabilities and perhaps even authorization) need to be conveyed through claims in the attestation. 
The difference between attester and wallet provider would be that if the wallet provider is in the iss, it needs to be entitled by the trusted list to issue it's own attestations whereas in the case of the attester, it's the attester who has that entitlement. 

Does this sound reasonable?

Reviewer 4: I'm in favor of having sub as optional and without any constraint about its value

the current proposal is concerning to me

Reviewer: in the call yesterday we came to the conclusion that the fundamental question is:
"Is the Wallet Provider always the Attester of the Wallet Instance?"

Indeed, we have already seen the reality that wallet providers may not have the capacity/willingness to run an attestation service on their own. And nobody will force an wallet provider to use a different attestation service. But we should keep this option open. Therefore I am voting to use:
- `iss` as the issuer of the wallet attestation
- `sub` as the client id
The version should not be referenced in there as this is responsibility of the wallet provider and not a burden to the issuer. If the Wallet Provider runs the attestation service on his own, we have two possibilities:
- a) `iss` equals `sub`
- b) `sub` may be omitted
I'm in favor of a)

Reviewer 2: @peppelinux making the `sub` optional cannot be done in HAIP and needs to be discussed in the IETF on the client attestation draft. For HAIP, I suggest we remove the concrete values and put in letter what the values are expected to be: code suggestion below.

Reviewer 4: in the implementation I'm working on the `sub` value is the thumbprint of the cnf.jwk since this latter uniquely identify, anyway, the wallet instance

>   "sub": "https://wallet.example.com",

Reviewer: ```suggestion
  "sub": "https://wallet.example.com/version/1.0.0",
```

Author: Can you please explain the rationale for the path extension?

Reviewer: With regard to eIDAS terms: iss describes the Wallet Solution Provider and sub descibes the specific Wallet Solution of a Wallet Instance

Reviewer 6: Hmm, the string `https://wallet.example.com/version/1.0.0` gives me the impression that there are THREE concepts. Namely,

1. [**per-application identifier**] The traditional client ID assigned to the wallet application. All versions and all per-device instances of the same wallet application use the same client ID.
2. [**per-version identifier**] The version of the wallet application. The string `https://wallet.example.com/version/1.0.0` implies this. The same version of the wallet application may be installed onto multiple devices.
3. [**per-device identifier**] Per-device instance of the wallet application. A key pair generated on the device is only for the per-device instance.

If these three concepts are competing for a single claim, `sub`, then reaching an agreement that satisfies everyone would be impossible. It would be necessary to create new claims as needed.

Reviewer 4: the version is not of interest for an issuer (nor a RP ...) since it should not collect all the changelog/release notes about all the wallet solutions, so that information is useless

Reviewer: please reject the version suggestion, see comment above

>   "iss": "https://wallet.example.com",
>   "sub": "https://wallet.example.com",

Reviewer 2: ```suggestion
  "iss": "<identifier of the issuer of this wallet attestation>",
  "sub": "<`client_id` of the OAuth client>",
```
@peppelinux

Reviewer 4: I would also like to add to this discussion that we agreed that the claim client_id represents the unique client identifier, even if there are divisions about who is the real client: the wallet solution or the wallet instance. I'm in favour of the latter, by having the distinction between `iss` and `sub` when the wallet instance is installed on a personal mobile device and all the respective crytpographic keys are stored by each material owner (wallet provider has its own, then the wallet instance has its own).

However, "client_id" is not always used to specify a client identifier since there are many way in OAuth2 to uniquely identify a client within a flow, below I mention for instance methods like:

- client_secret_basic
- client_secret_jwt
- private_key_jwt

even if these are not relevant for the wallet solution, I'd try to not mandate too many details within the specs at this current phase, by giving more flexibility in the material implementations on the way.

Author: incorporated Kristina's suggestion

### Wallet Attestation Schema {#wallet-attestation-schema}

Wallets MUST use attestations following the definition given in [@!I-D.ietf-oauth-attestation-based-client-auth].

In addition to this definition, the Wallet Attestation MAY contain the following claims in the `cnf` element:

> * `key_type`: OPTIONAL. JSON String that asserts the security mechanism the wallet uses to manage the private key associated with the public key given in the `cnf` claim. This mechanism is based on the capabilities of the execution environent of the wallet, this might be a secure element (in case of a wallet residing on a smartphone) or a Cloud-HSM (in case of a cloud wallet). This specification defines the following values for `key_type`:

Reviewer 4: ```suggestion
* `key_type`: OPTIONAL. JSON String that asserts the security mechanism the Wallet uses to manage the private key associated with the public key given in the `cnf` claim. This mechanism is based on the capabilities of the execution environent of the Wallet, this might be a secure element (in case of a wallet residing on a smartphone) or a Cloud-HSM (in case of a cloud Wallet). This specification defines the following values for `key_type`:
```

>   * `software`: It MUST be used when the wallet uses software-based key management.

Reviewer 4: ```suggestion
  * `software`: It MUST be used when the Wallet uses software-based key management.
```

>   * `hardware`: It MUST be used when the wallet uses software-based key management.

Reviewer 4: ```suggestion
  * `hardware`: It MUST be used when the wallet uses hardware-based key management.
```

>   * `hardware`: It MUST be used when the wallet uses software-based key management.

Reviewer 4: ```suggestion
  * `hardware`: It MUST be used when the Wallet uses software-based key management.
```

>   * `tee`: It SHOULD be used when the wallet uses the Trusted Execution Environment for key management.

Reviewer 4: ```suggestion
  * `tee`: It SHOULD be used when the Wallet uses the Trusted Execution Environment for key management.
```

  * `secure_enclave`: It SHOULD be used when the Wallet uses the Secure Enclave for key management.
  * `strong_box`: It SHOULD be used when the Wallet uses the Strongbox for key management.

>   * `secure_element`: It SHOULD be used when the Wallet uses the Trusted Execution Environment for key management.

Reviewer 4: definition about Secure Enclave, Strongbox and TEE are required if any external reference is given to the reader

what they are or where they are defined is an editorial requirement

>   * `hsm`: It SHOULD be used when the wallet uses Hardware Security Module (HSM).

Reviewer 4: ```suggestion
  * `hsm`: It SHOULD be used when the Wallet uses Hardware Security Module (HSM).
```

> * `user_authentication`: OPTIONAL. JSON String that asserts the security mechanism the wallet uses to authenticate access to the private key associated with the public key given in the `cnf` claim. This specification defines the following values for `user_authentication`: `System-Biometry`, `System-PIN`, `Internal-Biometry`, `Internal-PIN`, and `SecureElement-PIN`.

Reviewer 4: ```suggestion
* `user_authentication`: OPTIONAL. JSON String that asserts the security mechanism the Wallet uses to authenticate access to the private key associated with the public key given in the `cnf` claim. This specification defines the following values for `user_authentication`: `system_biometry`, `system_pin`, `internal_biometry`, `internal_pin`, and `secureelement_pin`.
```

to be then snake cased

The Wallet Attestation MAY also contain the following claim:

> * `aal`: OPTIONAL. JSON String asserting the authentication level of the wallet and the key as asserted in the `cnf` claim.

Reviewer 4: ```suggestion
* `aal`: OPTIONAL. JSON String asserting the authentication level of the Wallet and the key as asserted in the `cnf` claim.
```

To obtain the issuer's Public key for verification, wallet attestions MUST support web-based key resolution as defined in Section 5 of [@!I-D.terbu-sd-jwt-vc]. The JOSE header `kid` MUST be used to identify the respective key.

> This is an example of a wallet attestation:

Reviewer 4: ```suggestion
This is an example of a Wallet Instance Attestation:
```
